File: MockLibAndroid/SmartMockLib/src/main/java/com/crescentflare/smartmock/utility/SmartMockFileUtility.java
package com.crescentflare.smartmock.utility;

import android.content.Context;
import android.content.res.AssetFileDescriptor;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;

/**
 * Smart mock library utility: easily access files with file:/// or assets:/// prefix
 */
public class SmartMockFileUtility
{
    /**
     * Private constructor, only static methods allowed
     */

    private SmartMockFileUtility()
    {
    }


    /**
     * Utility functions
     */

    public static String[] list(Context context, String path)
    {
        String[] fileList = null;
        if (isAssetFile(path))
        {
            try
            {
                fileList = context.getAssets().list(getRawPath(path));
            }
            catch (IOException ignored)
            {
            }
        }
        else
        {
            fileList = new File(getRawPath(path)).list();
        }
        if (fileList != null)
        {
            List<String> filteredList = new ArrayList<>();
            for (String fileItem : fileList)
            {
                if (!fileItem.toLowerCase().equals("thumbs.db") && !fileItem.toLowerCase().equals(".ds_store"))
                {
                    filteredList.add(fileItem);
                }
            }
            return filteredList.toArray(new String[filteredList.size()]);
        }
        return null;
    }

    public static String[] recursiveList(Context context, String path)
    {
        String[] items = list(context, path);
        if (items != null)
        {
            List<String> files = new ArrayList<>();
            for (String item : items)
            {
                boolean isFile = SmartMockFileUtility.getLength(context, path + "/" + item) > 0;
                if (!isFile)
                {
                    String[] dirFiles = recursiveList(context, path + "/" + item);
                    if (dirFiles != null)
                    {
                        for (String dirFile : dirFiles)
                        {
                            files.add(item + "/" + dirFile);
                        }
                    }
                }
                else
                {
                    files.add(item);
                }
            }
            return files.toArray(new String[files.size()]);
        }
        return null;
    }

    public static InputStream open(Context context, String path)
    {
        if (isAssetFile(path))
        {
            try
            {
                return context.getAssets().open(getRawPath(path));
            }
            catch (IOException ignored)
            {
            }
        }
        else
        {
            try
            {
                new FileInputStream(getRawPath(path));
            }
            catch (FileNotFoundException ignored)
            {
            }
        }
        return null;
    }

    public static long getLength(Context context, String path)
    {
        if (isAssetFile(path))
        {
            try
            {
                AssetFileDescriptor info = context.getAssets().openFd(getRawPath(path));
                return info.getLength();
            }
            catch (IOException ignored)
            {
            }
            try
            {
                InputStream inputStream = context.getAssets().open(getRawPath(path));
                long length = inputStream.available();
                inputStream.close();
                return length;
            }
            catch (IOException ignored)
            {
            }
        }
        else
        {
            File file = new File(getRawPath(path));
            if (file.exists())
            {
                return file.length();
            }
        }
        return -1;
    }

    public static boolean exists(Context context, String path)
    {
        return getLength(context, path) >= 0;
    }

    public static String readFromInputStream(InputStream stream)
    {
        final int bufferSize = 1024;
        final char[] buffer = new char[bufferSize];
        final StringBuilder out = new StringBuilder();
        String result = null;
        try
        {
            Reader in = new InputStreamReader(stream, "UTF-8");
            for ( ; ; )
            {
                int rsz = in.read(buffer, 0, buffer.length);
                if (rsz < 0)
                {
                    break;
                }
                out.append(buffer, 0, rsz);
            }
            result = out.toString();
            stream.close();
        }
        catch (Exception ignored)
        {
        }
        return result;
    }

    public static String obtainMD5(Context context, String path)
    {
        InputStream inputStream = open(context, path);
        if (inputStream != null)
        {
            String result = "";
            MessageDigest digest;
            try
            {
                digest = MessageDigest.getInstance("MD5");
            }
            catch (NoSuchAlgorithmException exception)
            {
                return "";
            }
            byte[] buffer = new byte[8192];
            int read;
            try
            {
                while ((read = inputStream.read(buffer)) > 0)
                {
                    digest.update(buffer, 0, read);
                }
                byte[] md5sum = digest.digest();
                BigInteger bigInt = new BigInteger(1, md5sum);
                String output = bigInt.toString(16);
                result = String.format("%32s", output).replace(' ', '0');
                inputStream.close();
            }
            catch (IOException ignored)
            {
            }
            return result;
        }
        return "";
    }

    public static boolean isAssetFile(String path)
    {
        return path.startsWith("assets:///");
    }

    public static String getRawPath(String path)
    {
        if (isAssetFile(path))
        {
            return path.replace("assets:///", "");
        }
        return path.replace("file:///", "");
    }
}

File: MockLibIOS/SmartMockLib/Classes/utility/SmartMockFileUtility.swift
//
//  SmartMockFileUtility.swift
//  SmartMockServer Pod
//
//  Main library utility: easily access files with file:/// or bundle:/// or document:///
//

import CommonCrypto

class SmartMockFileUtility {
    
    // --
    // MARK: Initialization
    // --
    
    private init() {
        // Private constructor, only static methods allowed
    }


    // --
    // MARK: Utility functions
    // --
    
    static func list(fromPath: String) -> [String]? {
        if let fileList = try? FileManager.default.contentsOfDirectory(atPath: getRawPath(fromPath)) {
            var filteredList: [String] = []
            for fileItem in fileList {
                if fileItem.lowercased() != "thumbs.db" && fileItem.lowercased() != ".ds_store" {
                    filteredList.append(fileItem)
                }
            }
            return filteredList
        }
        return nil
    }
    
    static func recursiveList(fromPath: String) -> [String]? {
        if let items = list(fromPath: fromPath) {
            var files: [String] = []
            for item in items {
                let isFile = SmartMockFileUtility.getLength(ofPath: fromPath + "/" + item) > 0
                if !isFile {
                    if let dirFiles = recursiveList(fromPath: fromPath + "/" + item) {
                        for dirFile in dirFiles {
                            files.append(item + "/" + dirFile)
                        }
                    }
                } else {
                    files.append(item)
                }
            }
            return files
        }
        return nil
    }
    
    static func open(path: String) -> InputStream? {
        if let inputStream = InputStream(fileAtPath: getRawPath(path)) {
            inputStream.open()
            return inputStream
        }
        return nil
    }
    
    static func getLength(ofPath: String) -> Int {
        if let attr = try? FileManager.default.attributesOfItem(atPath: getRawPath(ofPath)) {
            if let fileType = attr[FileAttributeKey.type] {
                if fileType as? String == FileAttributeType.typeDirectory.rawValue {
                    return 0
                }
            }
            if let fileSize = attr[FileAttributeKey.size] {
                return (fileSize as! NSNumber).intValue
            }
            return 0
        }
        return -1
    }
    
    static func exists(path: String) -> Bool {
        if let _ = try? FileManager.default.attributesOfItem(atPath: getRawPath(path)) {
            return true
        }
        return false
    }
    
    static func readFromInputStream(_ stream: InputStream) -> String {
        let data = NSMutableData()
        var buffer = [UInt8](repeating: 0, count: 4096)
        while stream.hasBytesAvailable {
            let bytesRead = stream.read(&buffer, maxLength: buffer.count)
            data.append(buffer, length: bytesRead)
        }
        let result = String(data: data as Data, encoding: String.Encoding.utf8) ?? ""
        stream.close()
        return result
    }
    
    static func obtainSHA256(path: String) -> String {
        if let inputStream = open(path: path) {
            // Read stream into data
            let inputData = NSMutableData()
            var buffer = [UInt8](repeating: 0, count: 4096)
            while inputStream.hasBytesAvailable {
                let bytesRead = inputStream.read(&buffer, maxLength: buffer.count)
                inputData.append(buffer, length: bytesRead)
            }
            inputStream.close()

            // Obtain the SHA256 hash
            let data = inputData as Data
            let hash = data.withUnsafeBytes { (bytes: UnsafeRawBufferPointer) -> [UInt8] in
                var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
                CC_SHA256(bytes.baseAddress, CC_LONG(data.count), &hash)
                return hash
            }

            // Return SHA256 hash string formatted as hexadecimal
            return hash.map { String(format: "%02hhx", $0) }.joined()
        }
        return ""
    }
    
    static func getRawPath(_ path: String) -> String {
        if path.hasPrefix("bundle:///") {
            return (Bundle.main.resourcePath ?? "") + "/" + path.replacingOccurrences(of: "bundle:///", with: "")
        } else if path.hasPrefix("document:///") {
            let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
            let documentPath = paths[0]
            return documentPath + "/" + path.replacingOccurrences(of: "document:///", with: "")
        }
        return path.replacingOccurrences(of: "file:///", with: "/")
    }
    
}
